src/lib/meta/meta_blk_service.cpp

```diff
@@ -1074,85 +1081,165 @@ sisl::byte_array MetaBlkService::read_sub_sb_internal(const meta_blk* mblk) cons
-                                 mblk->hdr.h.type, ret, uint64_cast(mblk->hdr.h.compressed_sz),
-                                 uint64_cast(mblk->hdr.h.src_context_sz));
-                        HS_REL_ASSERT(false, "failed to decompress");
-                    } else {
-                        // decompressed_size must equal to input sz before compress
-                        HS_REL_ASSERT_EQ(uint64_cast(mblk->hdr.h.src_context_sz),
-                                         uint64_cast(decompressed_size)); /* since decompressed_size is >=0 it
-                                                                             is safe to cast to uint64_t */
-                        HS_LOG(DEBUG, metablk,
-                               "[type={}] Successfully decompressed, compressed_sz: {}, src_context_sz: {}, "
-                               "decompressed_size: {}",
-                               mblk->hdr.h.type, uint64_cast(mblk->hdr.h.compressed_sz),
-                               uint64_cast(mblk->hdr.h.src_context_sz), decompressed_size);
-                    }
+    meta_subtype_vec_t independent_subsystems;
+    std::unordered_set< meta_sub_type > visited;
+
+    for (const auto& [type, dependencies] : m_dependency_topological_graph) {
+        // get all the vertex whose indegree is not 0
+        for (const auto& dependency : dependencies) {
+            visited.insert(dependency);
+        }
+    }
+
+    for (const auto& [type, dependencies] : m_dependency_topological_graph) {
+        // outdegree == 0 and indegree == 0 means no dependency
+        if (dependencies.empty() && visited.find(type) == visited.end()) { independent_subsystems.push_back(type); }
+    }
+
+    for (const auto& independent_subsystem : independent_subsystems) {
+        m_dependency_topological_graph.erase(independent_subsystem);
+    }
+
+    // all the subsystems are divided into two parts.
+    // for subsystems in m_dependency_topological_graph, we need to do topological sort and recover in order.
+
+    const auto& dependent_subtype_recovery_order = topological_sort(m_dependency_topological_graph);
+    for (const auto& subtype : dependent_subtype_recovery_order) {
+        for (const auto& m : m_sub_info[subtype].meta_bids) {
+            auto mblk = m_meta_blks[m];
+            recover_meta_block(mblk);
+        }
 
-                    cb(mblk, decompressed_buf, mblk->hdr.h.src_context_sz);
+        if (do_comp_cb && m_sub_info[subtype].comp_cb) {
+            m_sub_info[subtype].comp_cb(true);
+            HS_LOG(DEBUG, metablk, "[type={}] completion callback sent.", subtype);
+        }
+    }
+
+    // for subsystems in independent_subsystems, we can use concurrent recovery if necessary.
+    // for now, we recover independent_subsystems in current thread.
+    // we can start a thread to recover independent_subsystems concurrently to improve performace in the future.
+    for (const auto& subtype : independent_subsystems) {
```

Comment:
no need to maintain independent_subsystems, use m_sub_info and deduce independent clients during register_handler.

Reply:
we can not do this， since during register_handler， a subsystem only know whether it depends on anyone, but it does not know anyone depends on it